File: benchmark/template/programs/cpp/micro/mb03.cpp
// MB-03 - Integer arithmetic: mixed add / xor / multiply-modulo / divide.
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>

struct Result {
    std::int64_t s_add;
    std::int64_t s_xor;
    std::int64_t s_mul;
    std::int64_t s_div;
};

static Result workload() {
    constexpr std::int64_t N = 120000000;

    std::int64_t x = 20263917;  // the seed itself is the initial generator state
    std::int64_t s_add = 0;
    std::int64_t s_xor = 0;
    std::int64_t s_mul = 1;
    std::int64_t s_div = 0;

    for (std::int64_t i = 0; i < N; ++i) {
        x = (48271 * x) % 2147483647;
        s_add = (s_add + x) % 2147483647;
        s_xor = s_xor ^ x;
        s_mul = (s_mul * 33 + (x % 97)) % 1000003;
        s_div = s_div + x / 1000;
    }

    return Result{s_add, s_xor, s_mul, s_div};
}

// Steady-mode anti-elimination sink. The workload body is a pure function of no
// arguments, so without an observable use of its result on every iteration clang
// treats the call as loop-invariant and sinks it out of the steady loop: verified
// on the frozen recipe, mb01 then reported 42 ns for six of seven iterations while
// the work ran once after the loop. One volatile store per iteration, outside
// every pinned computation, keeps each iteration doing the whole workload.
static volatile std::int64_t g_steady_sink;

int main(int argc, char** argv) {
    // Section 5.2 program modes: argv[1] in {once, steady}, argv[2] = K (default 7).
    const char* mode = argc > 1 ? argv[1] : "once";
    Result res{};
    if (std::strcmp(mode, "steady") == 0) {
        int K = argc > 2 ? std::atoi(argv[2]) : 7;
        if (K < 1) K = 7;
        for (int k = 0; k < K; ++k) {
            const auto t0 = std::chrono::steady_clock::now();
            res = workload();
            g_steady_sink = res.s_add;
            const auto t1 = std::chrono::steady_clock::now();
            std::printf("ITER %d %lld\n", k,
                        static_cast<long long>(
                            std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count()));
            std::fflush(stdout);
        }
    } else {
        res = workload();
    }

    std::printf("MB03 %lld %lld %lld %lld\n",
                static_cast<long long>(res.s_add), static_cast<long long>(res.s_xor),
                static_cast<long long>(res.s_mul), static_cast<long long>(res.s_div));
    return 0;
}

File: benchmark/template/programs/swift/micro/mb03.swift
// MB-03 - Integer arithmetic: mixed add / multiply / modulo / xor / divide.
#if canImport(Darwin)
import Darwin  // fflush(stdout) for the section 5.2 ITER lines
#else
import Glibc   // the same fflush(stdout) on Linux, where the benchmark runs
#endif

/// The whole workload body (methodology 06 section 4, MB-03).
func workload() -> String {
    let n = 120_000_000

    var x = 20263917
    var sAdd = 0
    var sXor = 0
    var sMul = 1
    var sDiv = 0

    for _ in 0..<n {
        x = (48271 * x) % 2147483647
        sAdd = (sAdd + x) % 2147483647
        sXor = sXor ^ x
        sMul = (sMul * 33 + (x % 97)) % 1000003
        sDiv = sDiv + x / 1000
    }

    return "MB03 \(sAdd) \(sXor) \(sMul) \(sDiv)"
}

// Section 5.2 program modes: argv[1] in {once, steady}, argv[2] = K (default 7).
func main() {
    let args = CommandLine.arguments
    let mode = args.count > 1 ? args[1] : "once"
    var line: String
    if mode == "steady" {
        var iterations = args.count > 2 ? (Int(args[2]) ?? 7) : 7
        if iterations < 1 {
            iterations = 7
        }
        let clock = ContinuousClock()
        line = ""
        for k in 0..<iterations {
            let elapsed = clock.measure {
                line = workload()
            }
            let parts = elapsed.components
            let ns = parts.seconds * 1_000_000_000 + parts.attoseconds / 1_000_000_000
            print("ITER \(k) \(ns)")
            fflush(stdout)  // section 5.2: each ITER line is flushed immediately
        }
    } else {
        line = workload()
    }
    print(line)
}

main()
